### ISSUE TITLE
ItemsRepeater.ItemTemplate does not work with RecyclingElementFactory or IElementFactory

### ISSUE THREAD

**Issue by zed220** [microsoft/microsoft-ui-xaml#3448]
**Describe the bug**
I try to use the RecyclingElementFactory or IElementFactory as an ItemsRepeater.ItemTemplate but fail - app crashes.
Doc here - [https://docs.microsoft.com/ru-ru/uwp/api/microsoft.ui.xaml.controls.itemsrepeater.itemtemplate?view=winui-2.4](https://docs.microsoft.com/ru-ru/uwp/api/microsoft.ui.xaml.controls.itemsrepeater.itemtemplate?view=winui-2.4)

**Steps to reproduce the bug**
1. Create xaml like following:
```
<ItemsRepeater ItemsSource="{x:Bind Items}">
	<ItemsRepeater.ItemTemplate>
		<RecyclingElementFactory/>
	</ItemsRepeater.ItemTemplate>
</ItemsRepeater>
```

**Expected behavior**
All works fine.

**Version Info**
NuGet package version: 
[Microsoft.WinUI 3.0.0-preview2.200713.0]

Windows app type:
| UWP              | Win32            |
| :--------------- | :--------------- |
| No | Yes |

| Windows 10 version                  | Saw the problem? |
| :--------------------------------- | :-------------------- |
| May 2020 Update (19041)            | Yes   |

| Device form factor | Saw the problem? |
| :----------------- | :--------------- |
| Desktop            | Yes |

**Additional context**
I have attached an example.
[App17.zip](https://github.com/microsoft/microsoft-ui-xaml/files/5406863/App17.zip)

**Comment by Felix-Dev** [comment]
Typically, you would provide a subclass of `RecyclingElementFactory` with a custom implementation by overriding the *ElementCore() methods like `GetElementCore(ElementFactoryGetArgs)`. When you just pass the base class, we currently throw an exception with message "Templates property cannot be null or empty."

Should we keep throwing an exception here, just show "nothing" or provide a basic Item.ToString() visualization?

Here is how we currently handle specifying just a base `DataTemplate` or `DataTemplateSelector`:
#### DataTemplate
```cs
public List<string> Data = new List<string>() { "Item 1", "Item 2" };
```

```xml
<muxc:ItemsRepeater ItemsSource="{x:Bind Data}">
    <muxc:ItemsRepeater.ItemTemplate>
        <DataTemplate/>
    </muxc:ItemsRepeater.ItemTemplate>
</muxc:ItemsRepeater>
```

Result: No crash and no Data item is visualized.

#### DataTemplateSelector
```xml
<muxc:ItemsRepeater ItemsSource="{x:Bind Data}">
    <muxc:ItemsRepeater.ItemTemplate>
        <DataTemplateSelector/>
    </muxc:ItemsRepeater.ItemTemplate>
</muxc:ItemsRepeater>
```

Result: App crash with XAMLParseException "No matching constructor found on type 'Windows.UI.Xaml.Controls.DataTemplateSelector'."

If I provide a custom `RecyclingElementFactory` implementation, yet only override either GetElementCore() or RecycleElementCore(), we crash too. For example
```cs
public class TestRecyclingElementFactory : RecyclingElementFactory
{
    protected override UIElement GetElementCore(Microsoft.UI.Xaml.Controls.ElementFactoryGetArgs args)
    {
        return new ListViewItem() { Content = "123" };
    }

    //protected override void RecycleElementCore(Microsoft.UI.Xaml.Controls.ElementFactoryRecycleArgs args)
    //{

    //}
}
```

crashes the app with Exception "System.ArgumentException: The parameter is incorrect." Removing the comment blocks around RecycleElementCore() will show two ListViewItems as expected. Perhaps we can improve this (i.e. implementation/documentation)?

On a related note, if I do provide a custom `RecyclingElementFactory` yet override GetElementCore() to return null, the app crashes as well:
```cs
protected override UIElement GetElementCore(Microsoft.UI.Xaml.Controls.ElementFactoryGetArgs args
{
    return null;
}
```

I think we should just "show nothing" (since UIElement = null) in this case instead of crashing.

@StephenLPeters @ranjeshj @chingucoding For your thoughts how to handle these cases.

**Comment by chingucoding** [comment]
It is not that easy to "show nothing", that's why we also throw an exception for null DataTemplates. One workaround we currently use in ItemTemplateWrapper is to render Rectangle elements for every data entry.

Regarding the RecycleFactory without recycle override, I think this fails since we create items to measure them for layouting and recycling them right away, that's why it doesn't just render but actually crash. I would argue that this is kind of by design, though crashing might not be ideal.

Regarding returning null for GetElementCore, I think ItemsRepeater is more "optimistic" and assumes that a factory doesn't return null which kind of makes sense since the default factory doesn't, and if you implement your own factory that returns null as element, it's a weird spot. It is not ideal agreed, but it is a very niche case, after all why would you implement an ElementFactory to have it return null.

In summary, the handling of null is not great, and could definitely be improved.

**Comment by ranjeshj** [comment]
@zed220 what is it you are trying to achieve by passing an empty data template ,selector or factory ? You can pass in a rectangle or something which is not visible which could achieve the same desired result perhaps ?

**Comment by zed220** [comment]
Hello,

I have created a simple project with overriding the GetElementCore and the RecycleElementCore and all works fine.

Am I right that the RecyclingElementFactory class can not work without custom implementation for
the GetElementCore and the RecycleElementCore methods?

As I can see, the RecyclingElementFactory class have the [complex and interesting implementation](https://github.com/microsoft/microsoft-ui-xaml/blob/master/dev/Repeater/RecyclingElementFactory.cpp) for these methods. Why I can't use it as default?

Many thanks!

**Comment by chingucoding** [comment]
Hi @zed220 , you can in fact use the RecyclingElementFactory in your app. For example, you can do the following:

```xaml
<muxc:ItemsRepeater ItemsSource="{x:Bind Data}" Layout="{StaticResource stackLayout}">
    <muxc:ItemsRepeater.ItemTemplate>
        <muxc:RecyclingElementFactory RecyclePool="{StaticResource recyclePool}">
            <DataTemplate x:Key="item">
                <TextBlock Text="{Binding}" />
            </DataTemplate>
        </muxc:RecyclingElementFactory>
    </muxc:ItemsRepeater.ItemTemplate>
</muxcs:ItemsRepeater>
```

**Comment by zed220** [comment]
Hi again,

All works as expected.

Can I use the x:Bind in the last example as following:

```
<DataTemplate x:Key="item" x:DataType="local:Item">
    <TextBlock Text="{x:Bind Value}" />
</DataTemplate>
```

I try to set it in my project but fail.

Thanks

**Comment by chingucoding** [comment]
Does your Item class have a `Value` property? Using x:Bind should work in a DataTemplate and is also used in the samples of the XAML Controls Gallery. If you are getting an exception or build errors, it would be helpful to share them so we can see what is causing the issue.

**Comment by zed220** [comment]
I have mistaken. All works as expected. For example:

```
<ScrollViewer>
	<ItemsRepeater ItemsSource="{x:Bind Items}">
		<ItemsRepeater.ItemTemplate>
			<RecyclingElementFactory>
				<DataTemplate x:Key="item" x:DataType="local:Item">
					<TextBlock Text="{x:Bind Value}"/>
				</DataTemplate>
				<RecyclingElementFactory.RecyclePool>
					<RecyclePool/>
				</RecyclingElementFactory.RecyclePool>
			</RecyclingElementFactory>
		</ItemsRepeater.ItemTemplate>
	</ItemsRepeater>
</ScrollViewer>
```

```
public class Item : INotifyPropertyChanged {
	public string Value { get; set; }

	public event PropertyChangedEventHandler PropertyChanged;
}
```

Many thanks!

**Comment by StephenLPeters** [comment]
@zed220 glad your issue was resolved, I'm going to close it. If something else comes up feel free to post back here.

**Comment by Felix-Dev** [comment]
@StephenLPeters Do we want to close this issue even though we still have all these app crashes? Ideally we can do something about that since app crashes really should be avoided whenever possible. Since we have already seen some discussion about them in this issue, the discussion, if desired, could be continued here.

**Comment by StephenLPeters** [comment]
Yeah, okay, my mistake.

**Comment by StephenLPeters** [comment]
Talking with @ranjeshj I think we actually do want to crash, I don't think there is a real scenario where the developer would be in this scenario outside of a test environment and it is easier to debug this crash than it is to debug an empty screen.

**Comment by ranjeshj** [comment]
If there are specific scenarios that are crashing unexpectedly, I think it makes sense to fix those. I think we discussed the scenarios in length in these two PRs.
https://github.com/microsoft/microsoft-ui-xaml/pull/1389
https://github.com/microsoft/microsoft-ui-xaml/pull/1358

Closing this bug for now. Please feel free to re-open if there is a specific scenario that is painful.

**Comment by knat** [comment]
[The doc](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.itemsrepeater.itemtemplate?view=winui-3.0) says ItemsRepeater.ItemTemplate can accept "a custom implementation of IElementFactory", when assign IElementFactory to ItemTemplate, exception was thrown:
```C#
    public class MyItemsRepeater : ItemsRepeater, IElementFactory
    {
        private readonly Stack<UIElement> _elements = new();
        private readonly Func<UIElement> _elementCreator;
        public MyItemsRepeater(Func<UIElement> elementCreator)
        {
            _elementCreator = elementCreator;
            ItemTemplate = this;//System.ArgumentException: 'Value does not fall within the expected range.'
        }
        UIElement IElementFactory.GetElement(ElementFactoryGetArgs args)
        {
            if (!_elements.TryPop(out var element)) element = _elementCreator();
            return element;
        }
        void IElementFactory.RecycleElement(ElementFactoryRecycleArgs args)
        {
            _elements.Push(args.Element);
        }
    }
```
Can you please tell me how to fix it?

**Comment by StephenLPeters** [comment]
@ranjeshj FYI

**Comment by sarosigy** [comment]
It is an issue in WinUI 2.7 also, and _needs_ to be solved.

The reason because the IElementFactory cannot be cast to IElementFactoryShim (is it an internal interface??)

And WinUI 2.7 **does not** have RecyclingElementFactory (still in preview since 2.5) -- making impossible to use _any_ ElementFactory!

**Comment by MuleaneEve** [comment]
I am working on a WinUI 3 app, and I was surprised to be stopped by this issue, even with the [WindowsAppSDK 1.1.2](https://www.nuget.org/packages/Microsoft.WindowsAppSDK/1.1.2).

I really need to be able to create elements programmatically.
